sdks/common/runner-api/src/main/proto/beam_runner_api.proto

```diff
@@ -0,0 +1,526 @@
+
+  // (Required) A payload determined by the URN, fully specifying this node.
+  google.protobuf.Any payload = 6;
+}
+
+// A PCollection!
+message PCollection {
+
+  // (Required) A pipeline-scoped unique id for this node.
+  int64 id = 1;
+
+  // (Required) A unique program-readable name for the value, such as
+  // "foo/baz/bar.out"
+  string label = 2;
+
+  // (Required) The id of the Coder for this PCollection.
+  int64 coder_ref = 3;
+
+  // (Required) Whether this PCollection is bounded or unbounded
+  IsBounded is_bounded = 4;
+
+  // (Required) The id of the windowing strategy for this PCollection.
+  int64 windowing_strategy_ref = 5;
+
+  // (Required) Static display data for this PCollection.
+  DisplayData display_data = 6;
+}
+
+// An applied PTransform! This does not contain the graph data, but only the
+// fields specific to a graph node that is a Runner API transform
+// between PCollections.
+message PTransform {
+
+  // (Required) A pipeline-scoped unique id for this node.
+  int64 id = 1;
+
+  // (Required) A unique name for the application node.
+  //
+  // This should be stable over multiple runs of a driver program for the
+  // purposes of logging and associating pipeline state with a node.
+  string label = 2;
+
+  // (Optional) A URN that identifies what type of transform this is.
+  //
+  // If absent, this must be an "anonymous" composite transform.
+  //
+  // If present, this URN should suffice to understand the payload and
+  // the two together should fully specify the SDK-independent and
+  // runner-independent semantics of the transform.
+  string urn = 3;
+
+  // (Optional) An arbitrary payload whose schema is determined by the URN.
+  //
+  // For primitive transform in the Runner API, the payload is well defined:
+  //
+  //  - when the URN is "urn:beam:transforms:pardo" it is a ParDoPayload
+  //  - when the URN is "urn:beam:transforms:read" it is a ReadPayload
+  //  - when the URN is "urn:beam:transforms:gbk" it is a GroupByKeyPayload
+  //  - when the URN is "urn:beam:transforms:window" it is a WindowPayload
+  //  - when the URN is "urn:beam:transforms:flatten" it is absent
+  //
+  // For some special composite transforms, the payload is also well-defined:
+  //
+  //  - when the URN is "urn:beam:transforms:combine" it is a CombinePayload
+  //
+  // For other composites, the payload will often be absent.
+  //
+  // TODO: duplicate PTransform's fields into each Payload to save indirection?
+  google.protobuf.Any payload = 4;
+
+  // (Required) Static display data for this PTransform application.
+  DisplayData display_data = 6;
+}
+
+// The payload for the primitive ParDo transform.
+message ParDoPayload {
+
+  // (Required) The DoFn for this ParDo.
+  FunctionSpec do_fn = 1;
+
+  // (Required) Additional pieces of context the DoFn may require that
+  // are not otherwise represented in the payload.
+  // (may force runners to execute the ParDo differently)
+  repeated DoFnContext required_context = 2;
```

Comment:
The same reason it is an enum - we envision perhaps a list of contextual requirements. Just an example we won't follow - today in the DataflowRunner we have USES_KEYED_STATE as well as REQUIRES_WINDOW_ACCESS.

I guess there is probably an idiomatic way to leave it singular and evolve it into repeating?

Two commits ago it was just a `bool requires_window_access` so I'm not terribly attached to it, except that the signature of the `DoFn` as it pertains to the runner needs to be indicated here.